include/core/CCompressedLfuCache.h

```diff
@@ -628,20 +652,24 @@ class CConcurrentCompressedLfuCache final
     bool guardRead(bool timeOut, TReadFunc func) const override {
         if (timeOut) {
             std::shared_lock<std::shared_timed_mutex> lock{m_Mutex, m_MaximumWait};
-            return func();
+            return lock.owns_lock() && func();
         }
         std::shared_lock<std::shared_timed_mutex> lock{m_Mutex};
         return func();
     }
 
-    void guardWrite(bool timeOut, TWriteFunc func) override {
+    bool guardWrite(bool timeOut, TWriteFunc func) override {
         if (timeOut) {
             std::unique_lock<std::shared_timed_mutex> lock{m_Mutex, m_MaximumWait};
-            func();
-            return;
+            if (lock.owns_lock()) {
+                func();
+                return true;
+            }
+            return false;
```

Comment:
I have a feeling the original code was unsafe if acquiring the mutex timed out, as it would just soldier on and call the write function regardless.  But if I was missing something and it was OK before then I'll revert this aspect of the changes.

Reply:
Whoops, just a complete oversight on my part.